### PR TITLE
feat(apigateway): grant methods to ApiKey

### LINKED ISSUE #8060: [apigateway] ApiKey should implement grant*() APIs
ApiKey Construct should implement IGrantable interface to grant Read write permissions to other constructs.

### Use Case

I usually create a number of microservices like stacks and inter-connect them using API gateway secured with API Key. 

Consider following scenario where `service A` and `service B` are microservices like stacks and so only interact with each other over HTTPS.
Now if `service B` needs to talk to `service A`, it will have to make requests to api exposed by `service A` with api key created in `service A`. 

Right now I do this by creating  
```typescript
      new PolicyStatement({
        effect: Effect.ALLOW,
        actions: ['apigateway:GET'],
        resources: [
          `arn:aws:apigateway:${this.region}::/apikeys/${apiKey.keyId}`,
        ],
      })
```
on Lambda role for  `serviceB`.

### Proposed Solution
Implement IGrantable interface to allow for `apiKey.grantRead(serviceHandler)`. This is much simpler to read and is commonly used throughout cdk. 
### Other

I can try implementing this feature if approved.😉️

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by ayush987goyal**
feat(apigateway): Add grant* methods to ApiKey

[Ref](https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-iam-policy-examples.html)

close #8060

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-apigateway/lib/api-key.ts [hunk 1]
```diff
@@ -112,6 +114,46 @@ export class ApiKey extends Resource implements IApiKey {
     });
 
     this.keyId = resource.ref;
+    this.keyArn = `arn:aws:apigateway:${Stack.of(this).region}::/apikeys/${this.keyId}`;
```
Comment (nija-at) [inline packages/@aws-cdk/aws-apigateway/lib/api-key.ts:117]: Use [`Stack.formatArn()`](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_core.Stack.html#format-wbr-arncomponents) instead.

File: packages/@aws-cdk/aws-apigateway/lib/api-key.ts [hunk 2]
```diff
@@ -112,6 +114,46 @@ export class ApiKey extends Resource implements IApiKey {
...
+  }
+
+  /**
+   * Permits the IAM principle all read operations through this key
```
Comment (nija-at) [inline packages/@aws-cdk/aws-apigateway/lib/api-key.ts:121]:
```suggestion
   * Permits the IAM principal all read operations through this key
```

File: packages/@aws-cdk/aws-apigateway/lib/api-key.ts [hunk 3]
```diff
@@ -112,6 +114,46 @@ export class ApiKey extends Resource implements IApiKey {
...
+  public grantRead(grantee: iam.IGrantable): iam.Grant {
+    return iam.Grant.addToPrincipal({
+      grantee,
+      actions: readPermissions,
+      resourceArns: [this.keyArn],
+    });
+  }
+
+  /**
+   * Permits the IAM principle all write operations through this key
```
Comment (nija-at) [inline packages/@aws-cdk/aws-apigateway/lib/api-key.ts:134]:
```suggestion
   * Permits the IAM principal all write operations through this key
```

**Review by nija-at — CHANGES_REQUESTED** [review]
Suggestion:

If you could implement these APIs on a class called `ApiKeyBase` and then have both `ApiKey` and the `Import` class extend this base class, then this functionality is also available to imported keys.

Something like this -

```ts
class ApiKeyBase extends Resource implements IApiKey {
   public grantRead(...)  { ... }
   public grantWrite(...) { ... } 
   ...
}

export class ApiKey extends ApiKeyBase {
  public static fromApiKeyId(scope: Construct, id: string, apiKeyId: string): IApiKey {
    class Import extends ApiKeyBase {
      public keyId = apiKeyId;
    }
    return new Import(scope, id);
  }
}
```

Note that, `ApiKeyBase` is not exported.

**Comment by ayush987goyal** [comment]
Hi @nija-at , you might have to add the pr-readme-exempt label to this one.

**Comment by nija-at** [comment]
> Hi @nija-at , you might have to add the pr-readme-exempt label to this one.

Instead, how about creating a new section called 'Api Key' at line 154 (and in the table of contents)? It feels like lines154-239 are talking about API keys. Then, add a couple of sentences around the grant APIs to that section

**Comment by nija-at** [comment]
I would prefer not to export `ApiKeyBase` and I think we don't need an `ApiKeyBaseProps`. 

I've pushed a commit that changes this and restructures the code so the export is not needed. Take a look and let me know what you think.

**Review by nija-at — CHANGES_REQUESTED** [review]

**Comment by ayush987goyal** [comment]
> I would prefer not to export ApiKeyBase and I think we don't need an ApiKeyBaseProps.

I've pushed a commit that changes this and restructures the code so the export is not needed. Take a look and let me know what you think.

These changes look good to me. Thanks!

**Review by nija-at — DISMISSED** [review]

**Review by nija-at — APPROVED** [review]

**Pull request merged**